### DISCUSSION TITLE
Multiple Re-renders and Reset not working

### DISCUSSION THREAD

**Discussion by tanetrail** [react-hook-form/react-hook-form#4191] [Q&A] [answered]
the below form is having multiple re-renders when ever a selection or input is provided and also reset is not working. Any idea where it is going wrong

https://stackblitz.com/edit/ionic-react-hook-form-rj1z46?file=src%2FApp.tsx

also trying to see if i can address the performance using FormProvider Performance with no luck
https://react-hook-form.com/advanced-usage#FormProviderPerformance

**Comment by bluebill1049** [comment]
can you share the steps the reproduce the issue?

**Reply by tanetrail** [reply]
https://stackblitz.com/edit/ionic-react-hook-form-rj1z46?file=src%2FApp.tsx

1. first time of the app load and after selecting a select option. the entire form is reloaded twice (console log has App, 2times select and input)
2. similar re-renders happens when we select the select options or input. But when we select the check box or range it will not re-render (may be because select and input are different components). Tried to use the FormProvider performance option to see if only the component that is changed will re-render not the entire app but no luck
3. Finally after providing all values Reset is not able to reset the values to blank

Trying to see if we can reduce the re-renders

**Reply by bluebill1049** [reply]
are you subscribe to any of the `formState`?

https://react-hook-form.com/api#formState

<img width="961" alt="Screen Shot 2021-02-15 at 9 36 35 am" src="https://user-images.githubusercontent.com/10513364/107891228-4dab6d80-6f71-11eb-8f0a-168c468bd954.png">

probably worth deconstructor those `formState`, instead pass the entire object down as prop.

**Reply by tanetrail** [reply]
I removed the formState completely on the above sample code(Not sure how to check the isDirty on the React.memo). Still seeing the entire app getting re-rendered when i choose a option in the select or type some input. wondering how can we render only the component that is changed and not the entire application.

![image](https://user-images.githubusercontent.com/61951630/107892411-8d7b4200-6eea-11eb-8e1d-04c5bf102a46.png)

**Reply by bluebill1049** [reply]
hmmm you want to use the latest version? see if this goes away?

**Reply by tanetrail** [reply]
Sure...just curios is that common in V6. like whenever there is a change in child component is it going to re-render the entire application and we cant reload only child not the parent every time. Thank you for your help

**Reply by bluebill1049** [reply]
That's not normal, Controller should isolate re-render.

**Comment by leapful** [comment, marked as answer]
@tanetrail @bluebill1049

- Reset button is not worked because prop "value" is not passed to IonSelect, adding "value" from Controller to IonSelect will fix the issue.

- Multiple re-rendering times are caused by method reset() inside onIonChange of Select.tsx

```
<IonSelect placeholder="Select One" value={value} onIonChange= {
                    e => {
                      onChange(e?.detail.value);
                   //   console.log("reset");
                   //   console.log(getValues());
                      reset({...getValues(), country: 'test' });
                    }
                  } >
```

=> Whenever the value of IonSelect component is changed, method reset() is triggered and the form state will be updated and forces re-rendering Select component ( which is a wrapper of Controller component).

Switching to setValue() will fix the issue: https://stackblitz.com/edit/ionic-react-hook-form-fft5nx?file=Select.tsx

NOTE: At the first time, when select value for Select, the re-rendering always happens because the dirty flag of form state is updated from true to false and it's an expected behavior.

**Reply by bluebill1049** [reply]
Thanks very much @leapful for the detailed investigation and solution.
